cmd/gpu_plugin/rm/gpu_plugin_resource_manager.go

```diff
@@ -0,0 +1,408 @@
+		envs:   envs,
+	}
+}
+
+type DeviceInfoMap map[string]*deviceInfo
+
+func NewDeviceInfoMap() DeviceInfoMap {
+	return DeviceInfoMap{}
+}
+
+func NewResourceManager() *ResourceManager {
+	clientset, err := getClientset()
+
+	if err != nil {
+		klog.Warning("no clientset available, resource manager not created, err:", err)
+		return nil
+	}
+
+	rm := ResourceManager{
+		nodeName:  os.Getenv("NODE_NAME"),
+		clientset: clientset,
+	}
+
+	klog.Info("GPU device plugin resource manager enabled")
+
+	return &rm
+}
+
+// ReallocateWithFractionalResources runs the fractional resource logic.
+// This intentionally only logs errors and returns with the UseDefaultAllocateMethodError,
+// in case any errors are hit. This is to avoid clusters filling up with unexpected admission errors.
+func (rm *ResourceManager) ReallocateWithFractionalResources(request *pluginapi.AllocateRequest) (*pluginapi.AllocateResponse, error) {
+	if !isInputOk(request) {
+		// it is better to leave allocated gpu devices as is and return
+		return nil, &dpapi.UseDefaultAllocateMethodError{}
+	}
+
+	podCandidate, err := rm.findAllocationPodCandidate()
+	if _, ok := err.(*retryErr); ok {
+		klog.Warning("retrying POD resolving after sleeping")
+		time.Sleep(retryTimeout)
+		podCandidate, err = rm.findAllocationPodCandidate()
+	}
+
+	if err != nil {
+		if _, ok := err.(*zeroPendingErr); !ok {
+			klog.Error("allocation candidate not found, perhaps the GPU scheduler extender is not called, err:", err)
+		}
+		// it is better to leave allocated gpu devices as is and return
+		return nil, &dpapi.UseDefaultAllocateMethodError{}
+	}
+
+	pod := podCandidate.pod
+	cards := containerCards(pod, podCandidate.allocatedContainers)
+	logPodInfo(pod, cards)
+
+	return rm.createAllocateResponse(cards)
+}
+
+func isInputOk(rqt *pluginapi.AllocateRequest) bool {
+	// so far kubelet calls allocate for each container separately. If that changes, we need to refine our logic.
+	if len(rqt.ContainerRequests) != 1 {
+		klog.Warning("multi-container allocation request")
+		return false
+	}
+
+	crqt := rqt.ContainerRequests[0]
+	for _, id := range crqt.DevicesIDs {
+		if id == monitorID {
+			return false // intentionally not printing anything
+		}
+	}
+
+	return true
+}
+
+// findAllocationPodCandidate tries to find the best allocation candidate pod, which must be:
+//  -pending for this node
+//  -using GPU resources in its spec
+//  -is found via grpc service with unallocated GPU devices
+// returns:
+//  -the candidate pod struct pointer and no error, or
+//  -errRetry if unsuccessful, but there is perhaps hope of trying again with better luck
+//  -errZeroPending if no pending pods exist anymore (which is fine)
+//  -any grpc communication errors
+func (rm *ResourceManager) findAllocationPodCandidate() (*podCandidate, error) {
+	// get map of pending pods for this node
+	pendingPods, err := rm.getNodePendingGPUPods()
+	if err != nil {
+		return nil, err
+	}
+
+	var candidate podCandidate
```

Comment:
This could be omitted.

Reply:
Ok I'll make an attempt getting rid of that. Not only are tests missing, all documentation is missing, lots of things like that. I'll get to work.